### PR TITLE
Batch entitiy start updates for remembered entities

### PR DESCRIPTION
**Pull request by chbatey**
Refs #29059

### CODE REVIEW

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala [hunk 1]
```diff
@@ -352,35 +358,51 @@ private[akka] class Shard(
...
           settings.tuningParameters.updatingStateTimeout)
 
-        context.become {
-          case RememberEntitiesShardStore.UpdateDone(`entityId`) =>
-            if (VerboseDebug) log.debug("Update of [{}] {} done", entityId, command)
+        context.become(waitingForUpdate(Map.empty))
+
+        def waitingForUpdate(pendingStarts: Map[EntityId, Option[ActorRef]]): Receive = {
+          // none of the current impls will send back a partial update, yet!
+          case RememberEntitiesShardStore.UpdateDone(ids) if ids == ids =>
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:365]: Perhaps the comparison is not worth it anymore when there is a set to compare with another set. I think I mostly put it there to detect bugs and it was cheap (if we really need to detect that we accidentally triggered two writes concurrently we could do that some other way?).

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala [hunk 2]
```diff
@@ -352,35 +358,51 @@ private[akka] class Shard(
           // and this could always fail before ddata store completes retrying writes
           settings.tuningParameters.updatingStateTimeout)
 
-        context.become {
-          case RememberEntitiesShardStore.UpdateDone(`entityId`) =>
-            if (VerboseDebug) log.debug("Update of [{}] {} done", entityId, command)
+        context.become(waitingForUpdate(Map.empty))
+
+        def waitingForUpdate(pendingStarts: Map[EntityId, Option[ActorRef]]): Receive = {
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:363]: Is it only starts?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala [hunk 3]
```diff
@@ -410,19 +432,32 @@ private[akka] class Shard(
...
   }
 
-  private def receiveStartEntity(start: ShardRegion.StartEntity): Unit = {
-    val requester = sender()
-    log.debug("Got a request from [{}] to start entity [{}] in shard [{}]", requester, start.entityId, shardId)
-    touchLastMessageTimestamp(start.entityId)
+  // this could be because of a start message or due to a new message for the entity
+  // if it is a start entity then start entity ack is sent after it is created
+  private def startEntities(entities: Map[EntityId, Option[ActorRef]]): Unit = {
+    val alreadyStarted = entities.filterKeys(entity => entityIds(entity))
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:438]: When are they already started?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala [hunk 4]
```diff
@@ -505,7 +540,7 @@ private[akka] class Shard(
         context.system.scheduler.scheduleOnce(entityRestartBackoff, self, RestartEntity(id))
       } else {
         // FIXME optional wait for completion as optimization where stops are not critical
-        waitForAsyncWrite(id, RememberEntitiesShardStore.RemoveEntity(id))(passivateCompleted)
+        waitForAsyncWrite(id, RememberEntitiesShardStore.RemoveEntity(id))(_ => passivateCompleted(id))
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:543]: If we always just ignore the id parameter for the lambda perhaps we shouldn't have it at all?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala [hunk 5]
```diff
@@ -30,7 +30,8 @@ import akka.persistence.SnapshotSelectionCriteria
 private[akka] final class EventSourcedRememberEntitiesProvider(typeName: String, settings: ClusterShardingSettings)
     extends RememberEntitiesProvider {
 
-  // this is backed by an actor using the same events as for state-store-mode=persistence
+  // this is backed by an actor using the same events, at the serailisation level, as the now removed PersistentShard when state-store-mode=persistence
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:33]: serialization

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/protobuf/ClusterShardingMessageSerializer.scala [hunk 6]
```diff
@@ -406,11 +404,14 @@ private[akka] class ClusterShardingMessageSerializer(val system: ExtendedActorSy
   private def entityStateFromBinary(bytes: Array[Byte]): EntityState =
     EntityState(sm.EntityState.parseFrom(bytes).getEntitiesList.asScala.toSet)
 
-  private def entityStartedToProto(evt: EntityStarted): sm.EntityStarted =
-    sm.EntityStarted.newBuilder().setEntityId(evt.entityId).build()
+  private def entityStartedFromBinary(bytes: Array[Byte]): EntitiesStarted =
+    EntitiesStarted(Set(sm.EntityStarted.parseFrom(bytes).getEntityId))
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/protobuf/ClusterShardingMessageSerializer.scala:408]: ah, we adapt here, clever, was wondering about that change from single to set

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala [hunk 7]
```diff
@@ -97,23 +97,23 @@ private[akka] final class EventSourcedRememberEntitiesStore(
   override def snapshotPluginId: String = settings.snapshotPluginId
 
   override def receiveRecover: Receive = {
-    case EntityStarted(id)                 => state = state.copy(state.entities + id)
+    case EntitiesStarted(ids)              => state = state.copy(state.entities ++ ids)
```
Comment (johanandren) [inline akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:100]: was there something about `Set.union` being much faster?

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala (hunk 1 above)
Comment (chbatey) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:365]: Okay i'll remove

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala (hunk 2 above)
Comment (chbatey) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:363]: Yes we don't batch removal of entities

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala (hunk 3 above)
Comment (chbatey) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:438]: When we receive a StartEntity but it's already been triggered by a message

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala (hunk 4 above)
Comment (chbatey) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/Shard.scala:543]: 👍

File: akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala (hunk 7 above)
Comment (chbatey) [reply akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/EventSourcedRememberEntities.scala:100]: It's impl is def union(that: GenSet[A]): This = this ++ that

**Comment by chbatey** [comment]
@johanandren if there's only small things in the PR do you mind if i follow up in the next PR? I've built on top of this one and will conflict myself

**Review by johanandren — APPROVED** [review]
Sounds good!

**Pull request merged**